agent/acs/handler/acs_handler.go

```diff
@@ -327,6 +273,102 @@ func startACSSession(ctx context.Context, client wsclient.ClientServer, timer tt
 	}
 }
 
+// applyBackoffAndGenerateReconnectAndDeregisterMessages handles errors from ACS. It
+// returns true if the session can reconnect to ACS. Else, it returns false. It also
+// applies the appropriate backoff policy for errors which can result in the handler
+// reconnecting to ACS. For an "InactiveInstance" error, it emits an event to the
+// deregister instance event stream.
+func (session *Session) applyBackoffAndGenerateReconnectAndDeregisterMessages(acsError error, connectToACS chan<- struct{}) bool {
+	if acsError == nil || acsError == io.EOF {
+		session.backoff.Reset()
+	} else if strings.HasPrefix(acsError.Error(), "InactiveInstanceException:") {
+		seelog.Debug("Container instance is deregistered, notifying listeners")
+		err := session.deregisterInstanceEventStream.WriteToEventStream(struct{}{})
```

Comment:
One concern with not retrying at all is the agent will not heal if the ECS server mistakenly vends an InactiveInstanceException.  How about retrying with a very long backoff in order to permit self-healing in this case?